backend/cfg/cfg_liveness.ml

```diff
@@ -45,58 +45,48 @@ struct
-        let across =
-          if Cfg.can_raise_basic instr.desc && instr.stack_offset > 0
-          then Reg.Set.union across exn.before
-          else across
-        in
-        let before = Reg.add_set_array across instr.arg in
-        { before; across }
-    | Reloadretaddr ->
-      { before = Reg.diff_set_array before Proc.destroyed_at_reloadretaddr;
-        across = Reg.Set.empty
-      }
-    | Pushtrap _ -> { before; across = before }
-    | Poptrap -> { before; across = before }
-    | Prologue -> { before; across = before }
+   fun ({ before; across = _ } as domain) ~exn instr ->
+    if Cfg.is_pure_basic instr.desc
+       && Reg.disjoint_set_array before instr.res
+       && (not (Proc.regs_are_volatile instr.arg))
+       && not (Proc.regs_are_volatile instr.res)
+    then
+      (* If the operation is without side-effects and the result is unused then
+         the arguments also aren't necessary. *)
+      { before; across = before }
+    else
+      instruction ~can_raise:(Cfg.can_raise_basic instr.desc) ~exn domain instr
 
   let terminator :
       domain -> exn:domain -> Cfg.terminator Cfg.instruction -> domain =
-   fun { before; across = _ } ~exn instr ->
+   fun domain ~exn instr ->
     match instr.desc with
     | Never -> assert false
-    | Always _ ->
-      { before = Reg.add_set_array before instr.arg; across = before }
-    | Parity_test _ ->
-      { before = Reg.add_set_array before instr.arg; across = before }
-    | Truth_test _ ->
-      { before = Reg.add_set_array before instr.arg; across = before }
-    | Float_test _ ->
-      { before = Reg.add_set_array before instr.arg; across = before }
-    | Int_test _ ->
-      { before = Reg.add_set_array before instr.arg; across = before }
-    | Switch _ ->
-      { before = Reg.add_set_array before instr.arg; across = before }
-    | Return -> { before = Reg.set_of_array instr.arg; across = Reg.Set.empty }
     | Tailcall (Self _) ->
-      { before = Reg.set_of_array instr.arg; across = Reg.Set.empty }
-    | Raise _ ->
-      { before = Reg.add_set_array exn.before instr.arg; across = exn.before }
-    | Tailcall (Func _) ->
-      { before = Reg.set_of_array instr.arg; across = Reg.Set.empty }
-    | Call_no_return _ ->
-      { before = Reg.add_set_array exn.before instr.arg; across = exn.before }
+      (* CR-someday azewierzejew: If the stamps for the tail call DomainState
+         argument and parameter were the same and Tailcall (Self _) had
+         [instr.arg = instr.reg] (either by removing the args or adding results
+         because currently there is nonempty array in args but empty in res)
+         then it would behave exactly the same as every other instruction. *)
+      (* We have to handle this as a special case because the registers for a
+         given DomainState argument and parameter pair have different stamps. *)
+      instruction
+        ~can_raise:(Cfg.can_raise_terminator instr.desc)
+        ~exn Domain.bot instr
```

Comment:
Shouldn't `Return` also be computed by passing `Domain.bot`?

Reply:
We will always have `Domain.bot` from definition of dataflow because `Return` has no successors.